data-prepper-plugins/saas-source-plugins/saas-crawler/src/main/java/org/opensearch/dataprepper/plugins/source/saas_crawler/coordination/scheduler/LeaderScheduler.java

```diff
@@ -0,0 +1,102 @@
+     */
+    private static final int DEFAULT_EXTEND_LEASE_MINUTES = 3;
+
+    /**
+     * Default interval to run lease check and shard discovery
+     */
+    private static final Duration DEFAULT_LEASE_INTERVAL = Duration.ofMinutes(1);
+
+    private final EnhancedSourceCoordinator coordinator;
+
+    private final Duration leaseInterval;
+
+    private LeaderPartition leaderPartition;
+
+    private final SaasSourcePlugin sourcePlugin;
+
+    private final Crawler crawler;
+
+    public LeaderScheduler(EnhancedSourceCoordinator coordinator,
+                           SaasSourcePlugin sourcePlugin,
+                           Crawler crawler) {
+        this.coordinator = coordinator;
+        this.leaseInterval =  DEFAULT_LEASE_INTERVAL;
+        this.sourcePlugin = sourcePlugin;
+        this.crawler = crawler;
+    }
+
+    @Override
+    public void run() {
+        LOG.debug("Starting Leader Scheduler for initialization and source partition discovery");
+
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                // Try to acquire the lease if not owned.
+                if (leaderPartition == null) {
+                    final Optional<EnhancedSourcePartition> sourcePartition = coordinator.acquireAvailablePartition(LeaderPartition.PARTITION_TYPE);
+                    if (sourcePartition.isPresent()) {
+                        LOG.info("Running as a LEADER node");
+                        leaderPartition = (LeaderPartition) sourcePartition.get();
+                    }
+                }
+                // Once owned, run Normal LEADER node process.
+                // May want to quit this scheduler if we don't want to monitor future changes
+                if (leaderPartition != null) {
+                    LeaderProgressState leaderProgressState = leaderPartition.getProgressState().get();
+                    long lastPollTime =  leaderProgressState.getLastPollTime();
+
+                    //Start crawling and create child partitions
+                    long updatedPollTime = crawler.crawl(lastPollTime, coordinator);
+                    leaderProgressState.setLastPollTime(updatedPollTime);
+                    leaderPartition.setLeaderProgressState(leaderProgressState);
+                    coordinator.saveProgressStateForPartition(leaderPartition, null);
+                }
+
+            } catch (Exception e) {
+                LOG.error("Exception occurred in primary scheduling loop", e);
+            } finally {
+                if(leaderPartition != null) {
+                    // Extend the timeout
+                    // will always be a leader until shutdown
+                    try {
+                        coordinator.saveProgressStateForPartition(leaderPartition, Duration.ofMinutes(DEFAULT_EXTEND_LEASE_MINUTES));
+                    } catch (final Exception e) {
+                        LOG.error("Failed to save Leader partition state. Retrying...");
```

Comment:
No ellipses in logs.

```suggestion
                        LOG.error("Failed to save Leader partition state. This process will retry.");
```

Reply:
modified